### PR TITLE
8291912: Introduce per-allocation target struct for members in G1PLABAllocator

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I get reviews for this change that refactors the many per-destination arrays in PLABAllocator into an array of struct as I promised in the review for [JDK-8288966](https://bugs.openjdk.org/browse/JDK-8288966)/https://github.com/openjdk/jdk/pull/9726? I think that makes the code easier to follow since there is less clutter.

Testing: tier1-5, no performance differences

Thanks,
Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291912](https://bugs.openjdk.org/browse/JDK-8291912): Introduce per-allocation target struct for members in G1PLABAllocator

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [54c93338](https://git.openjdk.org/jdk/pull/9959/files/54c93338323614001ae8c74b2147112b52f10834)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to [54c93338](https://git.openjdk.org/jdk/pull/9959/files/54c93338323614001ae8c74b2147112b52f10834)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9959/head:pull/9959` \
`$ git checkout pull/9959`

Update a local copy of the PR: \
`$ git checkout pull/9959` \
`$ git pull https://git.openjdk.org/jdk pull/9959/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9959`

View PR using the GUI difftool: \
`$ git pr show -t 9959`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9959.diff">https://git.openjdk.org/jdk/pull/9959.diff</a>

</details>

### CODE REVIEW

**Comment by tschatzl** [comment]
Ping for reviewers...

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1Allocator.hpp [hunk 1]
```diff
@@ -156,18 +156,33 @@ class G1PLABAllocator : public CHeapObj<mtGC> {
...
+    void initialize(uint num_alloc_buffers, size_t desired_plab_size, size_t tolerated_refills);
+
+    // Should we actually boost the PLAB size?
+    // The _plab_refill_counter reset value encodes the ResizePLAB flag value already, so no
+    // need to check here.
+    bool should_boost() const { return _plab_fill_counter == 0; }
+
+    void notify_plab_refill(size_t tolerated_refills, size_t next_plab_size);
+
+  } _data[G1HeapRegionAttr::Num];
```
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1Allocator.hpp:185]: What do you think about calling the struct just `PLABData` and the member `_destination_data`?

I think that would read a bit better where used.

**Review by kstefanj — APPROVED** [review]
Refactoring looks good, minor naming suggestion that you can do with as you like.

**Comment by tschatzl** [comment]
Thanks @walulyai @kstefanj for your reviews
/integrate

**Pull request closed without merging**